test/test_torch.py

```diff
@@ -6397,80 +6398,210 @@ def test_logical_and(self, device):
-        res2 = test_tens.clone()
-        for i in iter_indices(res2):
-            res2[i] = min(res2[i], max_val)
-        self.assertEqual(torch.isnan(res1), torch.isnan(res2))
+        m1_res = torch.clamp(m1, min=min_val)
+        self.assertEqual(torch.isnan(m1_base), torch.isnan(m1_res))
 
-        torch.clamp(test_tens, max=max_val, out=out)
-        self.assertEqual(torch.isnan(out), torch.isnan(res1))
+        out = torch.empty_like(m1)
+        torch.clamp(m1, min=min_val, out=out)
+        self.assertEqual(torch.isnan(m1_base), torch.isnan(out))
 
+    def test_clamp_max_propagates_nans(self, device):
+        max_val = 1
+        m1, m1_base = self.generate_clamp_baseline(device, max_vals=max_val, with_nans=True)
+
+        m1_res = torch.clamp(m1, max=max_val)
+        self.assertEqual(torch.isnan(m1_base), torch.isnan(m1_res))
+
+        out = torch.empty_like(m1)
+        torch.clamp(m1, max=max_val, out=out)
+        self.assertEqual(torch.isnan(m1_base), torch.isnan(out))
+
+    def test_clamp_raises_arg_errors(self, device):
+        m1 = torch.randn(100, dtype=torch.float, device=device)
         error_msg = 'At least one of \'min\' or \'max\' must not be None'
         with self.assertRaisesRegex(RuntimeError, error_msg):
             m1.clamp()
         with self.assertRaisesRegex(RuntimeError, error_msg):
             m1.clamp_()
+        with self.assertRaisesRegex(RuntimeError, error_msg):
+            torch.clamp(m1)
+
+    def test_clamp_with_tensors(self, device):
+        min_vals = torch.linspace(-1, 0, 100, device=device)  # [-1..0]
+        max_vals = torch.linspace(1, 0, 100, device=device)  # [1..0]
+        m1, m1_base = self.generate_clamp_baseline(device, min_vals, max_vals)
+
+        m1_res = torch.clamp(m1, min_vals, max_vals)
+        self.assertEqual(m1_base, m1_res)
+
+        out = torch.empty_like(m1)
+        torch.clamp(m1, min=min_vals, max=max_vals, out=out)
+        self.assertEqual(m1_base, out)
+
+        m1.clamp_(min_vals, max_vals)
+        self.assertEqual(m1_base, m1)
+
+    def test_clamp_with_tensors_min(self, device):
+        min_vals = torch.linspace(-1, 0, 100, device=device)  # [-1..0]
+        m1, m1_base = self.generate_clamp_baseline(device, min_vals=min_vals)
+
+        m1_res = torch.clamp(m1, min=min_vals)
+        self.assertEqual(m1_base, m1_res)
+
+        out = torch.empty_like(m1)
+        torch.clamp(m1, min=min_vals, out=out)
+        self.assertEqual(m1_base, out)
+
+    def test_clamp_with_tensors_max(self, device):
+        max_vals = torch.linspace(1, 0, 100, device=device)  # [1..0]
+        m1, m1_base = self.generate_clamp_baseline(device, max_vals=max_vals)
+
+        m1_res = torch.clamp(m1, max=max_vals)
+        self.assertEqual(m1_base, m1_res)
+
+        out = torch.empty_like(m1)
+        torch.clamp(m1, max=max_vals, out=out)
+        self.assertEqual(m1_base, out)
+
+    def test_clamp_with_tensors_propagates_nans(self, device):
```

Comment:
I really appreciate how thorough you're being but think if there's a way to roll these tests up that only takes a few lines.